File: app/src/main/java/com/v1rex/smartincubator/Activities/RegisterActivity.java
package com.v1rex.smartincubator.Activities;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.v1rex.smartincubator.R;

public class RegisterActivity extends AppCompatActivity {

    private static final String KEY_USERNAME_ENTRY = "email_entry";
    private static final String KEY_PASSWORD_ENTRY = "password_entry";

    private FirebaseAuth mAuth;

    private EditText mUserNameEditText, mPasswordEditText;
    private TextInputLayout mEmailTextInputLayout, mPasswordInputLayout;
    private Button mRegisterBtn;
    private LinearLayout mProgressRegisterLayout , mRegisterView;
    private TextView mLoginTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        // Getting instance of the firebase Auth for registering the User
        mAuth = FirebaseAuth.getInstance();

        mUserNameEditText = (EditText) findViewById(R.id.user_register_edit_text);
        mPasswordEditText = (EditText) findViewById(R.id.password_register_edit_text);
        mProgressRegisterLayout = (LinearLayout) findViewById(R.id.progess_register);
        mRegisterView = (LinearLayout) findViewById(R.id.registerView);
        mEmailTextInputLayout = (TextInputLayout) findViewById(R.id.input_layout_email_register);
        mPasswordInputLayout = (TextInputLayout) findViewById(R.id.input_layout_password_register);
        mLoginTextView = (TextView) findViewById(R.id.link_login);

        mLoginTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
                finish();
            }
        });

        mRegisterBtn = (Button) findViewById(R.id.register_action_btn);
        mRegisterBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                register();
            }
        });

        if (savedInstanceState != null){
            String savedEmail = savedInstanceState.getString(KEY_USERNAME_ENTRY);
            mUserNameEditText.setText(savedEmail);

            String savedPassword = savedInstanceState.getString(KEY_PASSWORD_ENTRY);
            mPasswordEditText.setText(savedPassword);
        }


    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {

        outState.putString(KEY_USERNAME_ENTRY, mUserNameEditText.getText().toString());
        outState.putString(KEY_PASSWORD_ENTRY, mPasswordEditText.getText().toString());

        super.onSaveInstanceState(outState);

    }

    @Override
    public void onBackPressed() {
        startActivity(new Intent(RegisterActivity.this, MainActivity.class));
        finish();
    }

    /**
     * Method for registering the user
     */

    private void register(){

        if(mAuth == null){
            return ;
        }

        // a boolean for canceling the registering if something is wrong
        boolean cancel = false;

        String email = mUserNameEditText.getText().toString();
        String password = mPasswordEditText.getText().toString();

        if (TextUtils.isEmpty(password) ) {
            mPasswordInputLayout.setError(getString(R.string.error_field_password_required));
            cancel = true;
        } else if(!isPasswordValid(password)){
            mPasswordInputLayout.setError(getString(R.string.error_invalid_password));
            cancel = true;
        }


        if (TextUtils.isEmpty(email)) {
            mEmailTextInputLayout.setError(getString(R.string.error_field_username_required));
            cancel = true;
        } else if (!isEmailValid(email)) {
            mEmailTextInputLayout.setError(getString(R.string.error_invalid_email));
            cancel = true;
        }

        // if everything is good , register the user
        if( cancel == false){
            mRegisterView.setVisibility(View.INVISIBLE);
            mProgressRegisterLayout.setVisibility(View.VISIBLE);

            mAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                @Override
                public void onComplete(@NonNull Task<AuthResult> task) {
                    mProgressRegisterLayout.setVisibility(View.GONE);
                    if(!task.isSuccessful()){
                        mUserNameEditText.setVisibility(View.VISIBLE);
                        mPasswordEditText.setVisibility(View.VISIBLE);
                        mRegisterBtn.setVisibility(View.VISIBLE);
                        mEmailTextInputLayout.setError(getString(R.string.error_action_registration_failed));
                    } else{
                        mRegisterView.setVisibility(View.GONE);
                        startActivity(new Intent(RegisterActivity.this, UserInformationsActivity.class));
                    }

                }
            });

        }
    }

    /**
     * Check if the email is good
     * @param email
     * @return boolean
     */
    private boolean isEmailValid(String email) {
        //TODO: Replace this with your own logic
        return email.contains("@");
    }

    /**
     * Check if the password is good
     * @param password
     * @return boolean
     */
    private boolean isPasswordValid(String password) {
        //TODO: Replace this with your own logic
        return password.length() > 4;
    }
}

File: app/src/main/java/com/v1rex/smartincubator/Activities/RegisterActivity.kt
package com.v1rex.smartincubator.Activities

import android.app.DatePickerDialog
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.support.design.widget.TextInputLayout
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.iid.FirebaseInstanceId
import com.v1rex.smartincubator.Model.User
import com.v1rex.smartincubator.R
import com.v1rex.smartincubator.R.drawable.calendar
import kotlinx.android.synthetic.main.activity_register.*
import kotlinx.android.synthetic.main.activity_send_messages.*
import java.text.SimpleDateFormat
import java.util.*

class RegisterActivity : AppCompatActivity() {

    private var mAuth: FirebaseAuth? = null
    private val database = FirebaseDatabase.getInstance()
    private val ref = database.getReference("Data")

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        // Getting instance of the firebase Auth for registering the User
        mAuth = FirebaseAuth.getInstance()

        link_login.setOnClickListener {
            startActivity(Intent(this@RegisterActivity, LoginActivity::class.java))
            finish()
        }

        register_action_btn.setOnClickListener { register() }

        if (savedInstanceState != null) {
            val savedEmail = savedInstanceState.getString(KEY_USERNAME_ENTRY)
            user_register_edit_text.setText(savedEmail)

            val savedPassword = savedInstanceState.getString(KEY_PASSWORD_ENTRY)
            password_register_edit_text.setText(savedPassword)
        }

        sexe_requierd_text_view.visibility = View.GONE
        account_type_requierd_text_view.visibility = View.GONE


        var cal = Calendar.getInstance()

        val dateSetLisener = DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
            cal.set(Calendar.YEAR, year)
            cal.set(Calendar.MONTH, monthOfYear)
            cal.set(Calendar.DAY_OF_MONTH, dayOfMonth)

            val myFormat = "dd.MM.yyyy" // mention the format you need
            val sdf : String = SimpleDateFormat(myFormat, Locale.US).format(cal.time)

            age_edit_text.setText(sdf)
        }

        age_edit_text.setOnClickListener {
            DatePickerDialog(this, dateSetLisener,
                    cal.get(Calendar.YEAR),
                    cal.get(Calendar.MONTH),
                    cal.get(Calendar.DAY_OF_MONTH)).show()
        }


    }

    /**
     * Method for registering the user
     */
    private fun register() {

        if (mAuth == null) {
            return
        }

        // a boolean for canceling the registering if something is wrong
        var cancel = false

        val email = user_register_edit_text.text.toString()
        val password = password_register_edit_text.text.toString()
        val lastName = last_name_edit_text.text.toString()
        val firstName = first_name_edit_text.text.toString()
        val age = age_edit_text.text.toString()
        var sexe = ""
        var accountType = ""

        val maleSelected = male_selected.isChecked
        val femaleSelected = female_selected.isChecked
        val startupSelected = startup_button.isChecked
        val mentorSlected = mentor_button.isChecked


        if(!isNetworkAvailable()){
            input_layout_password_register.error = getString(R.string.error_internet_required)
            cancel = true
        } else if (TextUtils.isEmpty(password)) {
            input_layout_password_register.error = getString(R.string.error_field_password_required)
            cancel = true
        } else if (!isPasswordValid(password)) {
            input_layout_password_register.error = getString(R.string.error_invalid_password)
            cancel = true
        }


        if (TextUtils.isEmpty(email)) {
            input_layout_email_register.error = getString(R.string.error_field_username_required)
            cancel = true
        } else if (!isEmailValid(email)) {
            input_layout_email_register.error = getString(R.string.error_invalid_email)
            cancel = true
        }


        if (TextUtils.isEmpty(lastName)) {
            last_name_input_text.error = getString(R.string.field_requierd)
            cancel = true
        }

        if (TextUtils.isEmpty(firstName)) {
            first_name_input_text.error = getString(R.string.field_requierd)
            cancel = true
        }

        if (TextUtils.isEmpty(age)) {
            age_input_text.error = getString(R.string.field_requierd)
            cancel = true
        }

        if (maleSelected == false && femaleSelected == false) {
            cancel = true
            sexe_requierd_text_view.visibility = View.VISIBLE

        } else {
            if (maleSelected == true) {
                sexe = "Male"

            } else {
                sexe = "Female"
            }

        }

        if (startupSelected == false && mentorSlected == false) {
            cancel = true
            account_type_requierd_text_view.visibility = View.VISIBLE
        } else {
            if (startupSelected == true) {
                accountType = "Startup"

            } else if (mentorSlected == true) {
                accountType = "Mentor"
            }

        }


        // if everything is good , register the user
        if (cancel == false) {
            registerView.visibility = View.INVISIBLE
            progess_register.visibility = View.VISIBLE

            mAuth!!.createUserWithEmailAndPassword(email, password).addOnCompleteListener(this) { task ->
                progess_register.visibility = View.GONE
                if (!task.isSuccessful) {
                    user_register_edit_text.visibility = View.VISIBLE
                    password_register_edit_text.visibility = View.VISIBLE
                    register_action_btn.visibility = View.VISIBLE
                    input_layout_email_register.error = getString(R.string.error_action_registration_failed)
                } else {

                    val user = User(lastName, firstName, email, sexe, age, accountType, mAuth!!.uid.toString())
                    var firebaseMessaging = FirebaseInstanceId.getInstance().token
                    user.registrationToken = firebaseMessaging.toString()
                    // Setting where to submit the user in the firebase realtime database
                    val usersRef = ref.child("users")
                    val userRef = usersRef.child(user.mUserId)
                    userRef.setValue(user)
                    registerView.visibility = View.GONE

                    if (mentorSlected == true) {
                    startActivity(Intent(this, MentorRegisterActivity::class.java))
                    } else if (startupSelected == true) {
                    startActivity(Intent(this, StartupRegisterActivity::class.java))
                    }

                }
            }

        }
    }

    private fun isNetworkAvailable(): Boolean {
        val connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetworkInfo = connectivityManager.activeNetworkInfo
        return activeNetworkInfo != null && activeNetworkInfo.isConnected
    }

    /**
     * Check if the email is good
     * @param email
     * @return boolean
     */
    private fun isEmailValid(email: String): Boolean {
        return email.toString().contains("@")
    }

    /**
     * Check if the password is good
     * @param password
     * @return boolean
     */
    private fun isPasswordValid(password: String): Boolean {
        return password.length > 4
    }

    companion object {
        private val KEY_USERNAME_ENTRY = "email_entry"
        private val KEY_PASSWORD_ENTRY = "password_entry"
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString(KEY_USERNAME_ENTRY, user_register_edit_text.text.toString())
        outState.putString(KEY_PASSWORD_ENTRY, password_register_edit_text.text.toString())
        super.onSaveInstanceState(outState)
    }

    override fun onBackPressed() {
        startActivity(Intent(this@RegisterActivity, LoginActivity::class.java))
        finish()
    }
}


